src/ray/core_worker/core_worker_process.h

```diff
@@ -141,13 +143,10 @@ class CoreWorkerProcessImpl {
   const CoreWorkerOptions options_;
 
   /// The core worker instance of this worker process.
-  std::shared_ptr<CoreWorker> core_worker_ ABSL_GUARDED_BY(mutex_);
```

Comment:
This is what I'm concerned about, thread annotation useful via providing compile-time check on lock acquisition / release precondition validation. While for `MutexProtected` I feel it's hard to apply annotation, which seems like a regression to me :(

Reply:
I think the new api is not possible to go wrong (read lock and call non cost methods), as long as it compiles. Whereas the old annotations can go wrong , so I don’t see regressions.

can you provide a code example that contains UB (or is just wrong), that could not compile with old api and can compile with new api?

Reply:
When you return a `WriteLocked` or `ReadLocked`, it's possible to forget acquiring mutex before accessing the value.

While for the current implementation, `ABSL_GUARDED_BY` makes sure access without mutex cannot compile.

Reply:
How can you forget to acquire a lock, since having a constructed ReadLocked means you already hold the reader lock?

Reply:
Let me clarify, I don't think there're anything wrong with your implementation;
but I feel it's better to apply thread annotation on getter functions and data members, especially the later, which provides compile-time check.

> How can you forget to acquire a lock, since having a constructed ReadLocked means you already hold the reader lock?

The assumption is you have a correct `ReadLocked` and `WriteLocked` here.

Reply:
I don't intend to block you from merging the PR, but you did loss the thread annotation, which was applied on every function and data member in the current impl.